_posts/2017-05-30-predicting-artist-career-stage-with-machine-learning-in-apache-spark.markdown

```diff
@@ -0,0 +1,181 @@
+
+<!-- more -->
+
+### Extracting Features
+
+First, we retrieve artist features, such as the top price fetched at auction and artist age along with manually assigned career stage value from HDFS using Hive. Here's a subset of the query.
+
+```scala
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.hive.HiveContext
+import org.joda.time.DateTime
+
+case class Features(
+  artistId:        String,
+  careerStageGene: Double,
+  age:             Int,
+  maxPrice:        Double
+)
+
+private def extractFeatures(hc: HiveContext): RDD[Features] = {
+  hc.sql(
+    s"""SELECT
+       |  artist_id,
+       |  career_stage_gene,
+       |  birthday,
+       |  max_price_in_auctions
+       |FROM
+       |  data.artists
+       |""".stripMargin
+  ).rdd.map { row =>
+    Features(
+      artistId        = row.getString(0),
+      careerStageGene = row.getDouble(1),
+      age             = DateTime.now.getYear - Try(DateTime.parse(row.getString(2)).getYear).getOrElse(DateTime.now.getYear),
+      maxPrice        = row.getDouble(3)
+    )
+  }
+}
+```
+
+Features are packed in a vector to become usable by the built-in Spark functions. We're not using the manually entered career stage gene as input, so it's not part of the feature vector.
+
+```scala
+import org.apache.spark.mllib.linalg._
+
+case class Features(
+   ...
+) = {
+  def vector: Vector = {
+    Vectors.dense(
+      age.toDouble,
+      maxPrice
+    )
+  }
+
+```
+
+### Normalizing and Weighing Features
+
+Since fetching a million dollars at auction isn't a million times more important than an artist's age, we must normalize values. We create a scaler to be used in both the training and the prediction parts of the code.
+
+```scala
+import org.apache.spark.mllib.feature.StandardScaler
+
+val scaler = new StandardScaler().fit(features.map(f => f.vector))
+```
+
+### Preparing Training Points
+
+Our training data contains all the artists that have a manual value for a career stage gene. The input is the scaled vector of features and the output is the known career stage value.
+
+```scala
+val featuresForTraining = features.filter(feature => feature.careerStageGene > 0)
+val labeledPointsForTraining = featuresForTraining.map {
+  f => LabeledPoint(f.careerStageGene, scaler.transform(f.vector))
+}
+```
+
+### Training a Model
+
+We use a linear regression with stochastic gradient descent and an intercept (the value of career stage when the input features are zero) and train it. The optimizer options below are the default, but may need some experimenting depending on input data.
```

Comment:
the mention of 'intercept' here is incidental to the subject of the sentence, as all linear models have an intercept. If you want to mention the role of the intercept, something like this will be better:

"We use a linear regression with stochastic gradient descent (SGD) as the learning scheme. This will yield coefficients for each feature along with an intercept (the value...)."